cores/esp8266/core_esp8266_waveform.cpp

```diff
@@ -173,137 +151,133 @@ int ICACHE_RAM_ATTR stopWaveform(uint8_t pin) {
+    waveformsToEnable = 0;
+    waveformsToDisable = 0;
     // Find the first GPIO being generated by checking GCC's find-first-set (returns 1 + the bit of the first 1 in an int32_t)
-    startPin = __builtin_ffs(waveformEnabled) - 1;
+    startPin = __builtin_ffs(waveformsEnabled) - 1;
     // Find the last bit by subtracting off GCC's count-leading-zeros (no offset in this one)
-    endPin = 32 - __builtin_clz(waveformEnabled);
+    endPin = 32 - __builtin_clz(waveformsEnabled);
   }
 
   bool done = false;
-  if (waveformEnabled) {
+  if (waveformsEnabled) {
+    uint32_t now = ESP.getCycleCount();
     do {
-      nextEventCycles = microsecondsToClockCycles(MAXIRQUS);
+      nextEventCycle = now + microsecondsToClockCycles(MAXIRQUS);
       for (int i = startPin; i <= endPin; i++) {
-        uint32_t mask = 1<<i;
+        uint32_t mask = 1UL<<i;
 
         // If it's not on, ignore!
-        if (!(waveformEnabled & mask)) {
+        if (!(waveformsEnabled & mask))
           continue;
-        }
 
-        Waveform *wave = &waveform[i];
-        uint32_t now = GetCycleCountIRQ();
+        Waveform *wave = &waveforms[i];
 
-        // Disable any waveforms that are done
-        if (wave->expiryCycle) {
-          int32_t expiryToGo = wave->expiryCycle - now;
-          if (expiryToGo < 0) {
-              // Done, remove!
-              waveformEnabled &= ~mask;
-              if (i == 16) {
-                GP16O &= ~1;
-              } else {
-                ClearGPIO(mask);
-              }
-              continue;
-            }
+        // Check for toggles etc.
+        if (ExpiryState::UPDATE == wave->hasExpiry) {
+          wave->expiryCycle += wave->nextServiceCycle; // in ExpiryState::UPDATE, expiryCycle temporarily holds relative cycle count
+          if (waveformsState & mask)
+            wave->expiryCycle += wave->nextTimeLowCycles; // update expiry time to next full period
+          wave->hasExpiry = ExpiryState::ON;
         }
-
-        // Check for toggles
-        int32_t cyclesToGo = wave->nextServiceCycle - now;
-        if (cyclesToGo < 0) {
-          waveformState ^= mask;
-          if (waveformState & mask) {
+        const int32_t cyclesToGo = wave->nextServiceCycle - now;
+        int32_t expiryToGo = (ExpiryState:: ON == wave->hasExpiry) ? wave->expiryCycle - now : (cyclesToGo + 1);
+        const int32_t nextEventCycles = nextEventCycle - now;
+        if (cyclesToGo >= expiryToGo) {
+          // don't update waveform on or after expiring
+          expiryToGo = 0;
+        }
+        else if (cyclesToGo > 0) {
+          if (nextEventCycles > cyclesToGo)
+            nextEventCycle = wave->nextServiceCycle;
+        }
+        else {
+          if (wave->nextTimeHighCycles && (!(waveformsState & mask) || !wave->nextTimeLowCycles)) {
+            waveformsState |= mask;
             if (i == 16) {
               GP16O |= 1; // GPIO16 write slow as it's RMW
-            } else {
+            }
```

Comment:
Same comment as the variable renaming.  Please try and follow whatever the prior code formatting/cuddling was.  I know we're not consistent, but when things like this get added it again distracts from the important logic changes.

Reply:
I am sorry, I had to completely refactor the identifiers to even begin to be able to get to where I needed to get. Before, I kept getting confused what with duty cycle/CPU cycle and so on.
I hope you can see things my way now - even at the expense that diff reading is futile now.
The main thing is, it's a major improvement for my use case, running Tone, servo control, and PWM on PC case fans all of a single ESP8266 at the same time.